ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockBroadcastValidatorImpl.java

```diff
@@ -0,0 +1,144 @@
+  private final SafeFuture<Boolean> consensusValidationSuccessResult;
+  private final SafeFuture<BroadcastValidationResult> broadcastValidationResult;
+
+  private BlockBroadcastValidatorImpl(
+      final BlockGossipValidator blockGossipValidator,
+      final BroadcastValidationLevel broadcastValidationLevel) {
+    this.blockGossipValidator = blockGossipValidator;
+    this.broadcastValidationLevel = broadcastValidationLevel;
+    consensusValidationSuccessResult = new SafeFuture<>();
+    broadcastValidationResult = new SafeFuture<>();
+  }
+
+  public static BlockBroadcastValidatorImpl create(
+      final SignedBeaconBlock block,
+      final BlockGossipValidator blockGossipValidator,
+      final BroadcastValidationLevel broadcastValidationLevel) {
+    BlockBroadcastValidatorImpl blockBroadcastValidator =
+        new BlockBroadcastValidatorImpl(blockGossipValidator, broadcastValidationLevel);
+    blockBroadcastValidator.buildValidationPipeline(block);
+    return blockBroadcastValidator;
+  }
+
+  @Override
+  public void onConsensusValidationSucceeded() {
+    consensusValidationSuccessResult.complete(true);
+  }
+
+  @Override
+  public void attachToBlockImport(final SafeFuture<BlockImportResult> blockImportResult) {
+    switch (broadcastValidationLevel) {
+      case NOT_REQUIRED, GOSSIP:
+        // GOSSIP validation isn't dependent on block import result,
+        // so not propagating exceptions to consensusValidationSuccessResult allow blocks\blobs
+        // to be published even in case block import fails before gossip validation completes
+        return;
+      case CONSENSUS, CONSENSUS_AND_EQUIVOCATION:
+        // Any successful block import will be considered as a consensus validation success, but
+        // more importantly we propagate exceptions to the consensus validation, thus we capture any
+        // early block import failures
+        blockImportResult
+            .thenApply(BlockImportResult::isSuccessful)
+            .propagateTo(consensusValidationSuccessResult);
+    }
+  }
+
+  @Override
+  public SafeFuture<BroadcastValidationResult> getResult() {
+    return broadcastValidationResult;
+  }
+
+  private void buildValidationPipeline(final SignedBeaconBlock block) {
+    // validateBroadcast should not be called at all but let's cover the case for safety
+    if (broadcastValidationLevel == NOT_REQUIRED) {
+      broadcastValidationResult.complete(SUCCESS);
+    }
+
+    // GOSSIP only validation
+    SafeFuture<BroadcastValidationResult> validationPipeline =
+        blockGossipValidator
+            .validate(block, true)
+            .thenApply(
+                gossipValidationResult -> {
+                  if (gossipValidationResult.isAccept()) {
+                    return SUCCESS;
+                  }
+                  return GOSSIP_FAILURE;
+                });
+
+    if (broadcastValidationLevel == GOSSIP) {
+      validationPipeline.propagateTo(broadcastValidationResult);
```

Comment:
Maybe I'm too paranoid but I'd also cancel or complete `consensusValidationSuccessResult` here and in NOT_REQUIRED case. Not sure though it will actually leaking

Reply:
I think it isn't necessary, an incomplete safe future won't leak. I can `cancel` for clarity though..